webnn/resources/utils.js

```diff
@@ -1,61 +1,307 @@
- * property in ``expected`` by the acceptable ULP distance ``nulp`` with given ``dataType`` data type.
- *
- * @param {string} op
- * @param {Array} actual - Array of test values.
- * @param {Array} expected - Array of values expected to be close to the values in ``actual``.
- * @param {number} [nulp=0] - A BigInt value indicates acceptable ULP distance, default 0.
- * @param {string} [dataType="float32"] - A data type string, default "float32",
- *     more data type strings, please see:
- *     https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+ * Run WebNN operation tests.
+ * @param {String} opName - An operation name
+ * @param {Array} tests - A test resources array
+ * @param {Object} buildGraphFunc - A function for building a graph
  */
-function assert_array_approx_equals_ulp(actual, expected, nulp, dataType)
-{
-  /*
-    * Test if two primitive arrays are equal within acceptable ULP distance
-    */
-  assert_true(actual.length === expected.length,
-              `assert_array_approx_equals_ulp actual length ${actual.length} should be equal to expected length ${expected.length}`);
-  let actualBitwise, expectedBitwise, distance;
-  for (let i = 0; i < actual.length; i++) {
+function testWebNNOperation(opName, tests, buildGraphFunc) {
+  /**
+   * Get precison tolerance value.
+   * @param {String} opName - An operation name
+   * @param {String} metricType - Value: 'ULP', 'ATOL'
+   * @param {String} operandType - Value: 'float32', 'float16', etc.,
+   *     more data type strings, please see:
+   *     https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+   * @param {Object} resources - Resources used for building a graph
+   * @returns {Number} A tolerance number
+   */
+  function getPrecisonTolerance(opName, metricType, operandType, resources) {
+    let tolerance;
+    if (metricType === 'ULP') {
+      switch (opName) {
+        case 'averagepool2d':
+          // TODO
+          break;
+        case 'conv2d':
+          // TODO
+          break;
+        case 'gemm':
+          // GEMM : alpha * (A x B) + beta * C
+          // An upper bound for the worst serial ordering is bounded by
+          // the number of lossy operations, where matrix multiplication
+          // is a dot product (mul and add times the number of elements)
+          // plus bias operations.
+          const options = resources.options;
+          const width = options.aTranspose ? shapeA[0] : shapeA[1];
+          tolerance = width * 2;
+          // default options.alpha is 1.0
+          if (options.alpha !== undefined && options.alpha !== 1.0) {
+            tolerance++;
+          }
+          if (options.c && options.beta !== 0.0) {
+            // default options.beta is 1.0
+            if (options.beta !== undefined && options.beta !== 1.0) {
+              tolerance++;
+            }
+            tolerance++;
+          }
+          break;
```

Comment:
@fdwr I created this `getPrecisonTolerance()` function, and add detail implementation to get a ULP tolerance for these operations which need compute by IEPOE, please take another look, thanks.

Reply:
Would it be simpler to just include the function name in the table, right beside all the other values, rather than needing to fill in a switch statement (because then there are two separate places to look for values and update, a table and a code switch)? e.g.

```
function getGemmPrecisionTolerance(shapeA, options) {
  const shapeA = resources.inputA.shape;
  options = resources.options;
  ...
}

...

const PrecisionMetrics = {
    batchNormalization: {ULP: {float32: 6, float16: 6}},
...
    gemm: {ULP: {float32: getGemmPrecisionTolerance, float16: getGemmPrecisionTolerance}}
...
    maxPool2d: {ULP: {float32: 0, float16: 0}},
...
};

...

function getPrecisonTolerance(opName, metricType, tensorDataType, resources) {
   let tolerance;
    if (metricType === 'ULP') {
      tolerance = PrecisionMetrics[opName].ULP[tensorDataType];
    }
    else if (metricType === 'ATOL') {
      tolerance = PrecisionMetrics[opName].ATOL[operandType];
    }

    // If the tolerance is dynamic, then evaluate the function to get the value.
    if (x instanceof Function) {
      tolerance = tolerance(resources);
    }
    return tolerance;
}
```

(but I'm also fine with the current code)

Reply:
Updated, thanks.